Cabal/Distribution/Lexer.x

```diff
@@ -0,0 +1,244 @@
+$instr           = [$printable $space] # [\"]
+$instresc        = $printable
+
+@nl          = \n | \r\n | \r
+@name        = $nameextra* $namecore $nameextra*
+@string      = \" ( $instr | \\ $instresc )* \"
+@numlike     = $digit [$digit \.]*
+@oplike      = [ \, \. \= \< \> \+ \* \- \& \| \! \$ \% \^ \@ \# \? \/ \\ \~ ]+
+
+tokens :-
+
+<0> {
+  $bom   ;
+  ()     ;
+}
+
+<bol_section, bol_field_layout, bol_field_braces> {
+  $spacetab* @nl                        { \_ _ _ -> adjustPos retPos >> lexToken }
+  $nbspspacetab+ @nl                    { \_ _ _ -> adjustPos retPos >> addWarning "Non-breaking space occured" >> lexToken }
+  -- no @nl here to allow for comments on last line of the file with no trailing \n
+  $spacetab* "--" $comment*             ;  -- TODO: check the lack of @nl works here
+                                        -- including counting line numbers
+}
+
+<bol_section> {
+  $spacetab*  --TODO prevent or record leading tabs
+                   { \pos len inp -> if B.length inp == len
+                                       then return (L pos EOF)
+                                       else setStartCode in_section
+                                         >> return (L pos (Indent len)) }
+  $spacetab* \{    { tok  OpenBrace }
+  $spacetab* \}    { tok  CloseBrace }
+}
+
+<in_section> {
+  $spacetab+   ; --TODO: don't allow tab as leading space
+
+  "--" $comment* ;
+
+  @name        { toki TokSym }
+  @string      { \p l i -> case reads (B.Char8.unpack (B.take l i)) of
+                             [(str,[])] -> return (L p (TokStr str))
+                             _          -> lexicalError p i }
+  @numlike     { toki TokNum }
+  @oplike      { toki TokOther }
+  $paren       { toki TokOther }
+  \:           { tok  Colon }
+  \{           { tok  OpenBrace }
+  \}           { tok  CloseBrace }
+  @nl          { \_ _ _ -> adjustPos retPos >> setStartCode bol_section >> lexToken }
+}
+
+<bol_field_layout> {
+  $spacetab*   --TODO prevent or record leading tabs
+                { \pos len inp -> if B.length inp == len
+                                    then return (L pos EOF)
+                                    else setStartCode in_field_layout
+                                      >> return (L pos (Indent len)) }
+}
+
+<in_field_layout> {
+  $spacetab+; --TODO prevent or record leading tabs
+  $field_layout' $field_layout*  { toki TokFieldLine }
+  @nl             { \_ _ _ -> adjustPos retPos >> setStartCode bol_field_layout >> lexToken }
+}
+
+<bol_field_braces> {
+   ()                { \_ _ _ -> setStartCode in_field_braces >> lexToken }
+}
+
+<in_field_braces> {
+  $spacetab+; --TODO prevent or record leading tabs
+  $field_braces' $field_braces*    { toki TokFieldLine }
+  \{                { tok  OpenBrace  }
+  \}                { tok  CloseBrace }
+  @nl               { \_ _ _ -> adjustPos retPos >> setStartCode bol_field_braces >> lexToken }
+}
+
+{
+
+data Token = TokSym   !ByteString
```

Comment:
Sigh... I would love docs, but I also understand there is a long and venerated tradition of totally underdocumented lexer implementations.